### PR TITLE
Issue/128 povgen

### PR DESCRIPTION
**Pull request by vitalych**

### CODE REVIEW

File: src/s2e/Plugins/VulnerabilityAnalysis/PovGenerationPolicy.cpp [hunk 1]
```diff
@@ -20,29 +20,35 @@ using namespace klee;
...
+
+    // Detecting vulnerability sources
+    m_process = s2e()->getPlugin<ProcessExecutionDetector>();
+    m_recipe = s2e()->getPlugin<recipe::Recipe>();
+    m_recipe->onPovReady.connect(sigc::bind(sigc::mem_fun(*this, &PovGenerationPolicy::onPovReadyHandler), false));
+
+    // Detecting platform
+    m_decreeMonitor = s2e()->getPlugin<DecreeMonitor>();
+    m_linuxMonitor = s2e()->getPlugin<LinuxMonitor>();
+    m_windowsCrashMonitor = s2e()->getPlugin<WindowsCrashMonitor>();
```
Comment (adrianherrera) [inline src/s2e/Plugins/VulnerabilityAnalysis/PovGenerationPolicy.cpp]: Is this used anywhere?

File: src/s2e/Plugins/VulnerabilityAnalysis/PovGenerator.cpp [hunk 2]
```diff
@@ -50,38 +37,70 @@ std::string PovGenerator::writeToFile(S2EExecutionState *state, const PovOptions
...
-    if (m_compress) {
-        povFileNameSS << ".zlib"; // use `zlib-flate -uncompress`
+    povFileNameSS << state->getID();
+
+    if (fileExtWithoutDot.size() > 0) {
+        povFileNameSS << "." << fileExtWithoutDot;
     }
 
-    std::string povFileName = g_s2e->getOutputFilename(povFileNameSS.str());
+    return g_s2e->getOutputFilename(povFileNameSS.str());
```
Comment (adrianherrera) [inline src/s2e/Plugins/VulnerabilityAnalysis/PovGenerator.cpp]: Can we use `s2e()->` instead? Using global vars like that is a little gross :)

File: src/s2e/Plugins/OSMonitors/Support/MemoryMap.cpp [hunk 3]
```diff
@@ -48,12 +55,147 @@ static MemoryMapRegionType WindowsProtectionToInternal(uint64_t protection) {
...
+    }
+
+    void removePid(uint64_t pid) {
+        MemoryMapRegionProcess::iterator it = m_regions.find(pid);
+        if (it != m_regions.end()) {
+            m_regions.erase(it);
+        }
+    }
+
+    void dump(llvm::raw_ostream &os) const {
```
Comment (adrianherrera) [inline src/s2e/Plugins/OSMonitors/Support/MemoryMap.cpp]: Again can we remove these `foreach2` macros?

**Comment by adrianherrera** [comment]
Wow, a lot of cool stuff in here.

Two things:

* Can we preference C++11 loops over the `foreach2` macro (which should really be deprecated). I noticed that there was one comment where you mentioned it didn't work for some reason, but it would be good to use them as much as possible.
* Can we also preference `s2e()` over `g_s2e` in plugin code - mainly for consistency and we can abstract over the global variable

File: src/s2e/Plugins/VulnerabilityAnalysis/PovGenerationPolicy.cpp (hunk 1 above)
Comment (vitalych) [reply src/s2e/Plugins/VulnerabilityAnalysis/PovGenerationPolicy.cpp]: nice catch!

File: src/s2e/Plugins/VulnerabilityAnalysis/PovGenerator.cpp (hunk 2 above)
Comment (vitalych) [reply src/s2e/Plugins/VulnerabilityAnalysis/PovGenerator.cpp]: yes

Comment (vitalych) [reply src/s2e/Plugins/VulnerabilityAnalysis/PovGenerator.cpp]: actually, no. These are all static functions, they don't have s2e() there.

File: src/s2e/Plugins/OSMonitors/Support/MemoryMap.cpp [hunk 4]
```diff
@@ -48,12 +55,147 @@ static MemoryMapRegionType WindowsProtectionToInternal(uint64_t protection) {
...
+    // This cannot be a non-static variable because it's used by the
+    // parent class but would be destroyed first, causing corruptions.
+    static Allocator s_alloc;
+
+public:
+    MemoryMapRegion() : IntervalMap(s_alloc) {
+    }
+
+    MemoryMapRegion(const MemoryMapRegion &other) : IntervalMap(s_alloc) {
+        foreach2 (it, other.begin(), other.end()) { insert(it.start(), it.stop(), *it); }
```
[the remark this answers is not in the document]
Comment (vitalych) [reply src/s2e/Plugins/OSMonitors/Support/MemoryMap.cpp]: It doesn't work, somehow the iterator becomes an integer in this case. Seems like the IntervalMap is not compatible with this construct.

File: src/s2e/Plugins/OSMonitors/Support/MemoryMap.cpp (hunk 3 above)
Comment (vitalych) [reply src/s2e/Plugins/OSMonitors/Support/MemoryMap.cpp]: I replaced them with the normal for(;;) style. It's almost as short as foreach2.

**Pull request merged**